### PR TITLE
Update lovefield definitions to v2.1.12.

### PR DESCRIPTION
**Pull request by freshp86**
Fixes https://github.com/google/lovefield/issues/212.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [lovefield.d.ts](https://github.com/google/lovefield/blob/4d5bb97f0a4aba9ddf9e2e2b23094a63561f54b9/dist/lovefield.d.ts#L69)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by freshp86** [comment]
I'll fix the lint errors and upload a new patch. Was having trouble running them locally.

**Comment by sandersn** [comment]
It's fine to update the definitions here, but if lovefield has its own, then maybe it would be better to remove the ones on DefinitelyTyped? [Instructions](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package)

**Comment by paulvanbrenk** [comment]
I agree with @sandersn here... with 2 sets of type definition, you'll always be chasing the implemtation. Closing.

**Comment by freshp86** [comment]
Even though Lovefield provides its own d.ts file, there are still people using it from here. I was thinking of updating this one to match the latest Lovefield version to accommodate such users. Looking at the instructions linked before, there is a proper way to remove definitions from DefinitelyTyped in a backwards compatible way, no?

So my proposal is to update these definitions to match Lovedfield's v2.1.12 and follow up with a deprecation per the instructions. Can we re-open?

**Comment by paulvanbrenk** [comment]
@freshp86 sounds good to me.

**Comment by freshp86** [comment]
Trying to run `npm run lint lovefield` locally, I keep getting the following error. Any ideas on how to fix this?

```
> definitely-typed@0.0.1 lint /usr/local/home/programs/open_source/DefinitelyTyped
> dtslint types "lovefield"

ts.readConfigFile is not a function

npm ERR! Linux 4.4.0-79-generic
npm ERR! argv "/usr/local/home/programs/node-v6.9.2-linux-x64/bin/node" "/usr/local/home/programs/node-v6.9.2-linux-x64/bin/npm" "run" "lint" "lovefield"
npm ERR! node v6.9.2
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! definitely-typed@0.0.1 lint: `dtslint types "lovefield"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the definitely-typed@0.0.1 lint script 'dtslint types "lovefield"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the definitely-typed package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     dtslint types "lovefield"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs definitely-typed
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls definitely-typed
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/local/home/programs/open_source/DefinitelyTyped/npm-debug.log
```

File: types/lovefield/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for Lovefield v2.0.62
+// Type definitions for Lovefield v2.1.12
```
Comment (paulvanbrenk) [inline types/lovefield/index.d.ts]: the version should be v2.1

**Comment by paulvanbrenk** [comment]
that's weird.. i had no trouble running on your repo. Did you do `npm install` in the root? Then you should be able to run `npm run lint lovefield` from the root.

**Comment by paulvanbrenk** [comment]
These are all the issues reported:

```
Testing lovefield
	Running: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/bin/tester/../../node_modules/dtslint/bin/index.js 
	/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/lovefield/index.d.ts
ERROR: 46:23   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 62:43   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 63:18   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 65:19   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 65:50   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 65:56   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 90:23   array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 102:20  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 102:38  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 138:15  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 159:17  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 209:34  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 209:48  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 211:28  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 213:20  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 213:34  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 215:40  array-type                       Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.
ERROR: 62:49   ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 65:62   ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 76:23   ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 78:18   ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 81:44   ban-types                        Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
ERROR: 83:46   ban-types                        Don't use 'Function' as a type. Avoid using the `Function` type. Prefer a specific function type, like `() => void`.
ERROR: 90:29   ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 136:26  ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 138:21  ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 172:24  ban-types                        Don't use 'Object' as a type. Avoid using the `Object` type. Did you mean `object`?
ERROR: 170:15  interface-name                   interface name must not have an "I" prefix
ERROR: 198:10  interface-over-type-literal      Use an interface instead of a type literal.
ERROR: 123:1   no-consecutive-blank-lines       Consecutive blank lines are forbidden
ERROR: 142:1   no-consecutive-blank-lines       Consecutive blank lines are forbidden
ERROR: 221:1   no-consecutive-blank-lines       Consecutive blank lines are forbidden
ERROR: 228:1   no-consecutive-blank-lines       Consecutive blank lines are forbidden
ERROR: 33:20   no-empty-interface               An empty interface is equivalent to `{}`.
ERROR: 34:20   no-empty-interface               An empty interface is equivalent to `{}`.
ERROR: 30:23   semicolon                        Missing semicolon
ERROR: 38:62   semicolon                        Missing semicolon
ERROR: 39:63   semicolon                        Missing semicolon
ERROR: 40:62   semicolon                        Missing semicolon
ERROR: 41:63   semicolon                        Missing semicolon
ERROR: 42:62   semicolon                        Missing semicolon
ERROR: 43:63   semicolon                        Missing semicolon
ERROR: 44:45   semicolon                        Missing semicolon
ERROR: 45:75   semicolon                        Missing semicolon
ERROR: 46:54   semicolon                        Missing semicolon
ERROR: 47:24   semicolon                        Missing semicolon
ERROR: 48:27   semicolon                        Missing semicolon
ERROR: 54:23   semicolon                        Missing semicolon
ERROR: 55:31   semicolon                        Missing semicolon
ERROR: 56:30   semicolon                        Missing semicolon
ERROR: 57:30   semicolon                        Missing semicolon
ERROR: 58:32   semicolon                        Missing semicolon
ERROR: 62:57   semicolon                        Missing semicolon
ERROR: 63:53   semicolon                        Missing semicolon
ERROR: 64:28   semicolon                        Missing semicolon
ERROR: 65:71   semicolon                        Missing semicolon
ERROR: 66:30   semicolon                        Missing semicolon
ERROR: 67:30   semicolon                        Missing semicolon
ERROR: 73:18   semicolon                        Missing semicolon
ERROR: 74:59   semicolon                        Missing semicolon
ERROR: 75:27   semicolon                        Missing semicolon
ERROR: 76:30   semicolon                        Missing semicolon
ERROR: 77:33   semicolon                        Missing semicolon
ERROR: 78:40   semicolon                        Missing semicolon
ERROR: 79:36   semicolon                        Missing semicolon
ERROR: 80:27   semicolon                        Missing semicolon
ERROR: 81:59   semicolon                        Missing semicolon
ERROR: 82:54   semicolon                        Missing semicolon
ERROR: 83:61   semicolon                        Missing semicolon
ERROR: 84:46   semicolon                        Missing semicolon
ERROR: 89:38   semicolon                        Missing semicolon
ERROR: 90:37   semicolon                        Missing semicolon
ERROR: 91:24   semicolon                        Missing semicolon
ERROR: 92:22   semicolon                        Missing semicolon
ERROR: 96:40   semicolon                        Missing semicolon
ERROR: 97:42   semicolon                        Missing semicolon
ERROR: 101:40  semicolon                        Missing semicolon
ERROR: 102:60  semicolon                        Missing semicolon
ERROR: 106:46  semicolon                        Missing semicolon
ERROR: 107:51  semicolon                        Missing semicolon
ERROR: 108:67  semicolon                        Missing semicolon
ERROR: 109:71  semicolon                        Missing semicolon
ERROR: 110:49  semicolon                        Missing semicolon
ERROR: 111:60  semicolon                        Missing semicolon
ERROR: 112:48  semicolon                        Missing semicolon
ERROR: 113:42  semicolon                        Missing semicolon
ERROR: 117:53  semicolon                        Missing semicolon
ERROR: 118:42  semicolon                        Missing semicolon
ERROR: 126:30  semicolon                        Missing semicolon
ERROR: 127:31  semicolon                        Missing semicolon
ERROR: 128:50  semicolon                        Missing semicolon
ERROR: 131:79  semicolon                        Missing semicolon
ERROR: 132:75  semicolon                        Missing semicolon
ERROR: 135:48  semicolon                        Missing semicolon
ERROR: 136:38  semicolon                        Missing semicolon
ERROR: 137:27  semicolon                        Missing semicolon
ERROR: 138:28  semicolon                        Missing semicolon
ERROR: 153:33  semicolon                        Missing semicolon
ERROR: 157:21  semicolon                        Missing semicolon
ERROR: 158:31  semicolon                        Missing semicolon
ERROR: 159:36  semicolon                        Missing semicolon
ERROR: 160:45  semicolon                        Missing semicolon
ERROR: 161:24  semicolon                        Missing semicolon
ERROR: 165:31  semicolon                        Missing semicolon
ERROR: 166:24  semicolon                        Missing semicolon
ERROR: 167:34  semicolon                        Missing semicolon
ERROR: 171:30  semicolon                        Missing semicolon
ERROR: 172:36  semicolon                        Missing semicolon
ERROR: 173:24  semicolon                        Missing semicolon
ERROR: 176:61  semicolon                        Missing semicolon
ERROR: 179:58  semicolon                        Missing semicolon
ERROR: 180:32  semicolon                        Missing semicolon
ERROR: 181:28  semicolon                        Missing semicolon
ERROR: 186:62  semicolon                        Missing semicolon
ERROR: 187:51  semicolon                        Missing semicolon
ERROR: 188:28  semicolon                        Missing semicolon
ERROR: 189:46  semicolon                        Missing semicolon
ERROR: 193:29  semicolon                        Missing semicolon
ERROR: 194:19  semicolon                        Missing semicolon
ERROR: 195:19  semicolon                        Missing semicolon
ERROR: 203:6   semicolon                        Missing semicolon
ERROR: 206:59  semicolon                        Missing semicolon
ERROR: 207:73  semicolon                        Missing semicolon
ERROR: 210:57  semicolon                        Missing semicolon
ERROR: 211:56  semicolon                        Missing semicolon
ERROR: 214:43  semicolon                        Missing semicolon
ERROR: 215:68  semicolon                        Missing semicolon
ERROR: 218:64  semicolon                        Missing semicolon
ERROR: 230:55  semicolon                        Missing semicolon
ERROR: 231:58  semicolon                        Missing semicolon
ERROR: 232:60  semicolon                        Missing semicolon
ERROR: 233:59  semicolon                        Missing semicolon
ERROR: 234:55  semicolon                        Missing semicolon
ERROR: 235:55  semicolon                        Missing semicolon
ERROR: 236:58  semicolon                        Missing semicolon
ERROR: 237:55  semicolon                        Missing semicolon
ERROR: 132:53  typedef-whitespace               expected onespace after colon in parameter
ERROR: 135:32  typedef-whitespace               expected nospace before colon in call-signature
ERROR: 135:25  typedef-whitespace               expected onespace after colon in parameter
ERROR: 132:53  whitespace                       missing whitespace
ERROR: 135:25  whitespace                       missing whitespace
ERROR: 242:1   export-just-namespace            Instead of `export =`-ing a namespace, use the body of the namespace as the module body.
ERROR: 121:3   no-padding                       Don't leave a blank line before '}'
ERROR: 240:1   no-padding                       Don't leave a blank line before '}'
ERROR: 7:3     strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 9:3     strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 19:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 24:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 29:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 33:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 34:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 37:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 53:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 61:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 70:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 72:3    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 88:5    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 95:5    strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 100:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 105:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 116:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 125:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 144:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 152:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 156:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 164:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 176:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 178:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 185:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 192:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 205:5   strict-export-declare-modifiers  'export' keyword is redundant here.
ERROR: 179:7   prefer-method-signature          Use a method signature instead of a property signature of function type.
ERROR: 1:25    dt-header                        Error parsing header. Expected: foo MAJOR.MINOR (patch version not allowed)
/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/lovefield/lovefield-tests.ts
ERROR: 25:11   object-literal-key-quotes        Unnecessarily quoted property 'id' found.
ERROR: 26:11   object-literal-key-quotes        Unnecessarily quoted property 'description' found.
ERROR: 27:11   object-literal-key-quotes        Unnecessarily quoted property 'deadline' found.
ERROR: 28:11   object-literal-key-quotes        Unnecessarily quoted property 'done' found.
ERROR: 21:7    only-arrow-functions             non-arrow functions are forbidden
ERROR: 32:7    only-arrow-functions             non-arrow functions are forbidden
ERROR: 36:7    only-arrow-functions             non-arrow functions are forbidden
ERROR: 37:25   only-arrow-functions             non-arrow functions are forbidden
ERROR: 44:7    only-arrow-functions             non-arrow functions are forbidden
ERROR: 46:15   only-arrow-functions             non-arrow functions are forbidden
ERROR: 48:15   only-arrow-functions             non-arrow functions are forbidden
ERROR: 19:4    semicolon                        Missing semicolon
ERROR: 49:43   semicolon                        Missing semicolon
ERROR: 39:21   whitespace                       missing whitespace
ERROR: 39:59   whitespace                       missing whitespace
ERROR: 3:3     no-var                           Do not use 'var'.
ERROR: 15:3    no-var                           Do not use 'var'.
ERROR: 16:3    no-var                           Do not use 'var'.
ERROR: 17:3    no-var                           Do not use 'var'.
ERROR: 24:9    no-var                           Do not use 'var'.
ERROR: 33:9    no-var                           Do not use 'var'.
```

**Comment by freshp86** [comment]
I am  able to run `npm test` succesfully, but not `npm run lint lovefield`. Anyway, I've fixed most of these errors locally, but some of them are not trivial to fix. Is it an option to override some of the tslint rules just for this subfolder?

**Comment by paulvanbrenk** [comment]
Yes, per the readme

>(To indicate that a lint rule truly does not apply, use `// tslint:disable rule-name` or better, `//tslint:disable-next-line rule-name.)`

**Comment by freshp86** [comment]
Is there anything left to be done here? I don't see any more errors.

**Comment by paulvanbrenk** [comment]
Instead of disabling the rules for the entire folder, can you be more granular and use the per line option?

**Comment by freshp86** [comment]
If disabling lint rules per line is desired, I'll address it hopefully soon. Ideally I would prefer not to spend more time on this PR, given that the plan is to eventually deprecate this extern file and rely only on the file that is distributed along with Lovefield itself.

**Pull request merged**